Zend/zend_exceptions.c

```diff
@@ -964,12 +970,13 @@ ZEND_API ZEND_COLD void zend_exception_error(zend_object *ex, int severity) /* {
 	ZVAL_OBJ(&exception, ex);
 	ce_exception = Z_OBJCE(exception);
 	EG(exception) = NULL;
-	if (ce_exception == zend_ce_parse_error) {
+	if (ce_exception == zend_ce_parse_error || ce_exception == zend_ce_compile_error) {
 		zend_string *message = zval_get_string(GET_PROPERTY(&exception, ZEND_STR_MESSAGE));
 		zend_string *file = zval_get_string(GET_PROPERTY_SILENT(&exception, ZEND_STR_FILE));
 		zend_long line = zval_get_long(GET_PROPERTY_SILENT(&exception, ZEND_STR_LINE));
 
-		zend_error_helper(E_PARSE, ZSTR_VAL(file), line, "%s", ZSTR_VAL(message));
+		zend_error_helper(ce_exception == zend_ce_parse_error ? E_PARSE : E_COMPILE_ERROR,
```

Comment:
There's a surprising consequence of this: CompileError gets treated like a "Fatal error" instead of a "Catchable fatal error" when running with `php -a`.

the CompileError thrown by token_get_all() also terminates `php -a`

Perhaps E_COMPILE_ERROR can be treated the same way E_PARSE is in the below block. I'm not sure how to differentiate zend_error_helper from zend_error_noreturn and other similar functions.

```c
// in main/main.c
			case E_ERROR:
			case E_CORE_ERROR:
			case E_COMPILE_ERROR:
			case E_USER_ERROR:
				error_type_str = "Fatal error";
				syslog_type_int = LOG_ERR;
				break;
```

```
php > throw new ParseError('x');
Parse error: x in php shell code on line 1
php > throw new CompileError('x');
Fatal error: x in php shell code on line 1
.... This terminates php -a

php > try {throw new CompileError('x');} catch (Throwable $_) { var_export($_); }
CompileError::__set_state(array(...
```